File: kotlin/java/com/google/ai/edge/litertlm/Message.kt
/*
 * Copyright 2025 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.google.ai.edge.litertlm

import com.google.gson.JsonArray
import com.google.gson.JsonObject
import kotlin.io.encoding.Base64
import kotlin.io.encoding.ExperimentalEncodingApi

/** The role of the message in a conversation. */
enum class Role(val value: String) {
  SYSTEM("system"), // represent the system
  USER("user"), // represent the user
  MODEL("model"), // represent the model
  TOOL("tool"), // represent a tool response
}

/**
 * Represents a message in the conversation. A message contains a [Role], a [Content] list, and an
 * optional list of [ToolCall].
 */
class Message
internal constructor(
  val role: Role,
  val contents: Contents = Contents.empty(),
  val toolCalls: List<ToolCall> = emptyList(),
  val channels: Map<String, String> = emptyMap(),
) {

  /** Convert to [JsonObject]. Used internally. */
  internal fun toJson() =
    JsonObject().apply {
      addProperty("role", role.value)
      if (contents.contents.isNotEmpty()) {
        add("content", contents.toJson())
      }
      if (toolCalls.isNotEmpty()) {
        val toolCallsJson = JsonArray()
        for (toolCall in toolCalls) {
          toolCallsJson.add(toolCall.toJson())
        }
        add("tool_calls", toolCallsJson)
      }
      if (channels.isNotEmpty()) {
        val channelsJson = JsonObject()
        for ((key, value) in channels) {
          channelsJson.addProperty(key, value)
        }
        add("channels", channelsJson)
      }
    }

  /** Convert the message to a string. */
  override fun toString() = contents.toString()

  companion object {

    /** Creates a system [Message] from the given text. */
    fun system(text: String) = system(Contents.of(text))

    /** Creates a system [Message] from the given contents. */
    fun system(contents: Contents) = Message(Role.SYSTEM, contents)

    /** Creates a user [Message] from the given text. */
    fun user(text: String) = user(Contents.of(text))

    /** Creates a user [Message] from the given contents. */
    fun user(contents: Contents) = Message(Role.USER, contents)

    /** Creates a model [Message] from the given text. */
    fun model(text: String) = model(Contents.of(text))

    /** Creates a model [Message] from the given contents, tool calls, and channels. */
    fun model(
      contents: Contents = Contents.empty(),
      toolCalls: List<ToolCall> = emptyList(),
      channels: Map<String, String> = emptyMap(),
    ) = Message(Role.MODEL, contents, toolCalls, channels)

    /** Creates a tool [Message] from the given contents. */
    fun tool(contents: Contents) = Message(Role.TOOL, contents)

    /** Creates a user [Message] from a text string. */
    @Deprecated("Use factory methods like user(), model() or Contents.of().")
    fun of(text: String) = user(text)

    /** Creates a user [Message] from the array of [Content]. */
    @Deprecated("Use factory methods like user(), model() or Contents.of().")
    fun of(vararg contents: Content) = user(Contents.of(contents.toList()))

    /** Creates a user [Message] from a list of [Content]. */
    @Deprecated("Use factory methods like user(), model() or Contents.of().")
    fun of(contents: List<Content>) = user(Contents.of(contents))
  }
}

class Contents private constructor(val contents: List<Content>) {
  /** Convert to [JsonObject]. Used internally. */
  internal fun toJson() =
    JsonArray().apply {
      for (content in contents) {
        this.add(content.toJson())
      }
    }

  /** Convert the Contents to a string. */
  override fun toString() = contents.joinToString("")

  companion object {

    /** Creates an empty [Contents] list. */
    internal fun empty() = Contents(emptyList())

    /** Creates a [Contents] from a text string. */
    fun of(text: String) = Contents.of(Content.Text(text))

    /** Creates a [Contents] from the array of [Content]. */
    fun of(vararg contents: Content) = Contents.of(contents.toList())

    /** Creates a [Contents] from a list of [Content]. */
    fun of(contents: List<Content>) = Contents(contents)
  }
}

/** Tool call returned by the model. ToolCalls are stored in [Message.toolCalls]. */
data class ToolCall(val name: String, val arguments: Map<String, Any?>) {
  internal fun toJson() =
    JsonObject().apply {
      addProperty("type", "function")
      val functionObj =
        JsonObject().apply {
          addProperty("name", name)
          add("arguments", arguments.toJsonObject())
        }
      add("function", functionObj)
    }
}

/** Represents a content in the [Message] of the conversation. */
sealed class Content {
  /** Convert to [JsonObject]. Used internally. */
  internal abstract fun toJson(): JsonObject

  /** Text. */
  data class Text(val text: String) : Content() {
    override fun toJson() =
      JsonObject().apply {
        addProperty("type", "text")
        addProperty("text", text)
      }

    override fun toString() = text
  }

  /** Image provided as raw bytes. */
  @OptIn(ExperimentalEncodingApi::class)
  data class ImageBytes(val bytes: ByteArray) : Content() {
    override fun toJson() =
      JsonObject().apply {
        addProperty("type", "image")
        addProperty("blob", Base64.encode(bytes))
      }
  }

  /** Image provided by a file. */
  data class ImageFile(val absolutePath: String) : Content() {
    override fun toJson() =
      JsonObject().apply {
        addProperty("type", "image")
        addProperty("path", absolutePath)
      }
  }

  /** Audio provided as raw bytes. */
  @OptIn(ExperimentalEncodingApi::class)
  data class AudioBytes(val bytes: ByteArray) : Content() {
    override fun toJson() =
      JsonObject().apply {
        addProperty("type", "audio")
        addProperty("blob", Base64.encode(bytes))
      }
  }

  /** Audio provided by a file. */
  data class AudioFile(val absolutePath: String) : Content() {
    override fun toJson() =
      JsonObject().apply {
        addProperty("type", "audio")
        addProperty("path", absolutePath)
      }
  }

  /** Tool response provided by the user when automatic tool calling is disabled. */
  data class ToolResponse(val name: String, val response: Any?) : Content() {
    override fun toJson() =
      JsonObject().apply {
        addProperty("type", "tool_response")
        addProperty("name", name)
        add("response", response.toJsonElement())
      }
  }
}

File: swift/Message.swift
// Copyright 2026 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import Foundation
import OSLog

/// The role of the message in a conversation.
public enum Role: String {
  case system
  case user
  case model
  case tool
}

/// Represents a content in the `Message` of the conversation.
///
/// Example Usage:
/// ```swift
/// let textContent = Content.text("What's the capital of France?")
/// let imageContent = Content.imageData(imageData)
/// let audioContent = Content.audioFile(path: "/path/to/audio.wav")
/// ```
public enum Content {
  /// Text.
  case text(String)
  /// Image provided as raw bytes.
  case imageData(Data)
  /// Image provided by an absolute file path.
  case imageFile(String)
  /// Audio provided as raw bytes.
  case audioData(Data)
  /// Audio provided by an absolute file path.
  case audioFile(String)
  /// Tool response.
  case toolResponse(name: String, response: Any, id: String = "")

  /// Convert to JSON format. Used internally.
  var toJson: [String: Any] {
    switch self {
    case .text(let text):
      return ["type": "text", "text": text]
    case .imageData(let bytes):
      return ["type": "image", "blob": bytes.base64EncodedString()]
    case .imageFile(let absPath):
      return ["type": "image", "path": absPath]
    case .audioData(let bytes):
      return ["type": "audio", "blob": bytes.base64EncodedString()]
    case .audioFile(let absPath):
      return ["type": "audio", "path": absPath]
    case .toolResponse(let name, let response, let id):
      var dict: [String: Any] = [
        "type": "tool_response",
        "name": name,
        "response": response,
      ]
      if !id.isEmpty {
        dict["id"] = id
      }
      return dict
    }
  }
}

/// Represents a message in the conversation. A message can contain multiple `Content`s.
///
/// Example Usage:
/// ```swift
/// let textMessage = Message("What's the capital of France?")
/// let modelMessage = Message("The capital of France is Paris.", role: .model)
/// let audioContentMessage = Message(of: someAudioContent)
/// let multiContentMessage = Message(of: someText, someImageContent)
/// let multiContentMessageFromArray = Message(contents: [someText, someImageContent])
/// ```
public struct Message {

  private let logger = Logger(
    subsystem: "com.google.odml.litertlm.swift",
    category: "Message"
  )

  /// The role of the message.
  public let role: Role

  /// The contents of the message.
  public let contents: Contents

  /// The channels of the message.
  public let channels: [String: String]

  /// The tool calls in the message.
  public let toolCalls: [ToolCall]

  // TODO: (b/459796564): Update constructor to `Message.role("messageTextHere")`
  /// Creates a `Message` from a text string.
  /// - Parameter text: The text content of the message.
  public init(_ text: String, role: Role = .user, channels: [String: String] = [:]) {
    self.init(
      contents: Contents(contents: [.text(text)]), role: role, channels: channels, toolCalls: [])
  }

  /// Creates a `Message` from one or more `Content`s.
  /// - Parameter contents: The list of contents for this message.
  public init(of contents: Content..., role: Role = .user) {
    precondition(!contents.isEmpty, "Contents should not be empty.")
    self.contents = Contents(contents: contents)
    self.role = role
    self.channels = [:]
    self.toolCalls = []
  }

  /// Creates a `Message` from a list of `Content`s.
  /// - Parameter contents: The list of contents for this message.
  public init(
    contents: [Content], role: Role = .user, channels: [String: String] = [:],
    toolCalls: [ToolCall] = []
  ) {
    precondition(
      !contents.isEmpty || !channels.isEmpty || !toolCalls.isEmpty,
      "Contents, channels and toolCalls should not all be empty.")
    self.contents = Contents(contents: contents)
    self.role = role
    self.channels = channels
    self.toolCalls = toolCalls
  }

  /// Creates a `Message` from a `Contents` object.
  /// - Parameter contents: The contents object for this message.
  public init(
    contents: Contents, role: Role = .user, channels: [String: String] = [:],
    toolCalls: [ToolCall] = []
  ) {
    precondition(
      !contents.isEmpty || !channels.isEmpty || !toolCalls.isEmpty,
      "Contents, channels and toolCalls should not all be empty.")
    self.contents = contents
    self.role = role
    self.channels = channels
    self.toolCalls = toolCalls
  }

  /// Convert to JSON format. Used internally.
  var toJson: [String: Any] {
    var dict: [String: Any] = ["role": role.rawValue]
    if !contents.isEmpty {
      dict["content"] = contents.toJson
    }
    if !channels.isEmpty {
      dict["channels"] = channels
    }
    if !toolCalls.isEmpty {
      dict["tool_calls"] = toolCalls.map { toolCall in
        let tcDict: [String: Any] = [
          "id": toolCall.id,
          "type": "function",
          "function": [
            "name": toolCall.name,
            "arguments": toolCall.arguments,
          ],
        ]
        return tcDict
      }
    }
    return dict
  }

  /// Convenience property to get the text content of the message.
  /// Returns a string by concatenating all text content items, separated by a space.
  /// Returns an empty string if no text content exists.
  public var toString: String {
    return contents.toString
  }

  /// A computed property that returns the JSON string representation
  /// of the Message, or nil if conversion fails.
  var jsonString: String {
    get throws {
      let jsonData = try JSONSerialization.data(withJSONObject: self.toJson, options: [])
      guard let resultString = String(data: jsonData, encoding: .utf8) else {
        throw LiteRTLMError.message(.failedToConvertToJson)
      }
      return resultString
    }
  }
}

/// Represents a collection of `Content` items in the `Message` of the conversation.
///
/// Example Usage:
/// ```swift
/// let contents = Contents(contents: [.text("Hello")])
/// ```
public struct Contents: RandomAccessCollection {
  public typealias Element = Content
  public typealias Index = Int

  /// The list of underlying content items.
  public var contents: [Content]

  public var startIndex: Int { contents.startIndex }
  public var endIndex: Int { contents.endIndex }

  public subscript(position: Int) -> Content {
    return contents[position]
  }

  /// Creates a `Contents` from a list of `Content`s.
  public init(contents: [Content]) {
    self.contents = contents
  }

  /// Creates an empty `Contents`.
  public static func empty() -> Contents {
    return Contents(contents: [])
  }

  /// Creates a `Contents` from a text string.
  public static func of(_ text: String) -> Contents {
    return Contents(contents: [.text(text)])
  }

  /// Creates a `Contents` from one or more `Content`s.
  public static func of(_ contents: Content...) -> Contents {
    return Contents(contents: contents)
  }

  /// Creates a `Contents` from a list of `Content`s.
  public static func of(_ contents: [Content]) -> Contents {
    return Contents(contents: contents)
  }

  /// Convert to JSON format. Used internally.
  var toJson: [[String: Any]] {
    return contents.map { $0.toJson }
  }

  /// A computed property that returns the JSON string representation
  /// of the Contents array.
  var jsonString: String {
    get throws {
      let jsonData = try JSONSerialization.data(withJSONObject: self.toJson, options: [])
      guard let resultString = String(data: jsonData, encoding: .utf8) else {
        throw LiteRTLMError.message(.failedToConvertToJson)
      }
      return resultString
    }
  }

  /// Returns a string by concatenating all text content items, separated by a space.
  public var toString: String {
    contents.compactMap { content in
      if case .text(let str) = content { return str } else { return nil }
    }.joined(separator: " ")
  }
}

/// Represents a tool call from the model.
public struct ToolCall {
  public let name: String
  public let id: String
  public let arguments: [String: Any]

  public init(name: String, id: String, arguments: [String: Any]) {
    self.name = name
    self.id = id
    self.arguments = arguments
  }
}
